### PR TITLE
"view item details"

### PR DESCRIPTION
**Pull request by romatowskak**

### CODE REVIEW

File: src/app/action-items/action-items.component.css [hunk 1]
```diff
@@ -103,6 +103,11 @@
     border-bottom: 2px solid #3d7dca;
 }
 
+.router-link-active {
+    background-color: rgb(200, 224, 233) !important;
```
Comment (mdybich) [inline src/app/action-items/action-items.component.css]: Is there any reason to use important here?

File: src/app/action-items/action-items.component.ts [hunk 2]
```diff
@@ -16,36 +19,43 @@ export interface ActionItemMapped extends ActionItem {
...
     private daysCountService: DaysLeftToDeadlineService,
-    private matDialog: MatDialog
+    private matDialog: MatDialog,
+    private route: ActivatedRoute,
+    private router: Router,
+    private daysLeftPipe: DaysLeftCountedPipe
   ) {}
   ngOnInit() {
     this.retrieveActionItems();
+    this.route.queryParams.subscribe(params => {
```
Comment (mdybich) [inline src/app/action-items/action-items.component.ts]: Please add unsubscribe handling. Also it would be nice to move it to the separate method.

File: src/app/action-items/action-items.component.ts [hunk 3]
```diff
@@ -16,36 +19,43 @@ export interface ActionItemMapped extends ActionItem {
...
       .getAllItems()
       .pipe(
         first(),
         map(items => {
           const mappedActionItems: ActionItemMapped[] = items.map(item => {
-            const dueDayCounted = item.dueDate
-              ? this.daysCountService.daysLeftToDeadline(item.dueDate, this.currentDate)
-              : undefined;
-            return { ...item, dueDay: dueDayCounted };
+            return this.daysLeftPipe.transform(item, item.dueDate, this.currentDate);
```
Comment (mdybich) [inline src/app/action-items/action-items.component.ts]: I think we can use the pipe directly in the template.

File: src/app/action-items/action-items.component.ts [hunk 4]
```diff
@@ -60,4 +70,11 @@ export class ActionItemsComponent implements OnInit {
         }
       });
   }
+  private getActionItem(): void {
+    this.actionItem = this.tasksService.getActionItem(this.actionItemId);
+  }
+
+  changePath() {
```
Comment (mdybich) [inline src/app/action-items/action-items.component.ts]: Could you explain this logic? I think it is not needed since redirecting to the default `/items` ulr is handled in routing configuration.

File: src/app/item-details/item-details.component.css [hunk 5]
```diff
@@ -1,5 +1,116 @@
 .selectItemParagraph {
     text-align: center;
-    font-size: 13px;
+    font-size: 14px;
     color: #979797;
+    margin-bottom: 50px;
+}
+
+.details {
+    text-align: left;
```
Comment (mdybich) [inline src/app/item-details/item-details.component.css]: This style is not needed.

File: src/app/item-details/item-details.component.ts [hunk 6]
```diff
@@ -1,15 +1,28 @@
...
-  ngOnInit() {
+export class ItemDetailsComponent implements OnChanges {
+  @Output() changePath = new EventEmitter();
+  @Input() item: ActionItem;
+  @Input() id: string;
+  @Input() isLoadingActionItem: boolean;
+  @Input() currentDate: Date;
+  constructor(private daysLeftPipe: DaysLeftCountedPipe) {}
+  ngOnChanges() {
+    this.isLoadingActionItem = !this.isLoadingActionItem;
```
Comment (mdybich) [inline src/app/item-details/item-details.component.ts]: FIX: You should not modify `Input` values. I think that this logic should be moved to "container".

File: src/app/item-details/item-details.component.ts [hunk 7]
```diff
@@ -1,15 +1,28 @@
...
+    if (this.item) {
+      this.isLoadingActionItem = false;
+      this.item = this.daysLeftPipe.transform(this.item, this.item.dueDate, this.currentDate);
```
Comment (mdybich) [inline src/app/item-details/item-details.component.ts]: Why not to use pipe inside template?

File: src/app/item-details/item-details.component.ts [hunk 8]
```diff
@@ -1,15 +1,28 @@
...
+    }
+    if (this.item === undefined) {
```
Comment (mdybich) [inline src/app/item-details/item-details.component.ts]: I believe this if statement is not needed.

File: src/app/pipes/daysLeftCountedPipe/days-left-counted.pipe.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
+import { Pipe, PipeTransform } from '@angular/core';
+import { DaysLeftToDeadlineService } from 'src/app/services/daysLeftToDeadlineService/days-left-to-deadline.service';
+import { ActionItem } from 'src/app/services/tasksService/tasks.service';
+
+@Pipe({
+  name: 'daysLeftCounted'
+})
+export class DaysLeftCountedPipe implements PipeTransform {
+  constructor(private daysCountService: DaysLeftToDeadlineService) {}
+  transform(value: any, dueDate: Date, currentDate: Date): ActionItem {
```
Comment (mdybich) [inline src/app/pipes/daysLeftCountedPipe/days-left-counted.pipe.ts]: I think that this pipe should return only number value with days left value.

File: tsconfig.json [hunk 10]
```diff
@@ -1,7 +1,7 @@
 {
   "compileOnSave": false,
   "compilerOptions": {
-    "strictNullChecks": true,
+    "strictNullChecks": false,
```
Comment (mdybich) [inline tsconfig.json]: FIX: Please revert this change.

File: src/app/add-item/add-item.component.ts [hunk 11]
```diff
@@ -53,14 +51,16 @@ export class AddItemComponent implements OnInit {
...
+    const itemId = parseInt(this.tasksService.getLastItemId()) + 1;
     const newItem: ActionItem = {
-      title: this.dialogForm.get('name').value,
+      title: this.dialogForm.get('name')!.value,
       projectName: projectNameValue.name,
       type: 'General',
       completed: '60',
-      dueDate: !!dueDateValue ? dueDateValue : undefined
+      dueDate: !!dueDateValue ? dueDateValue : undefined,
+      id: JSON.stringify(itemId)
```
Comment (odudycz) [inline src/app/add-item/add-item.component.ts]: **SUGGESTION:** You can use `itemId.toString()`

File: src/app/pipes/daysLeftCountedPipe/days-left-counted.pipe.ts [hunk 12]
```diff
@@ -0,0 +1,19 @@
...
+import { DaysLeftToDeadlineService } from 'src/app/services/daysLeftToDeadlineService/days-left-to-deadline.service';
+
+@Pipe({
+  name: 'daysLeftCounted'
+})
+export class DaysLeftCountedPipe implements PipeTransform {
+  private currentDate: Date = new Date();
+  constructor(private daysCountService: DaysLeftToDeadlineService) {}
+  transform(value: ActionItem): number | undefined {
+    if (value) {
```
Comment (odudycz) [inline src/app/pipes/daysLeftCountedPipe/days-left-counted.pipe.ts]: **FIX:** Please invert if to reduce nesting:
```
if(!value) {
  return;
}
const dueDayCounted = value.dueDate;
(...)
```

File: src/app/services/tasksService/tasks.service.ts [hunk 13]
```diff
@@ -80,12 +89,31 @@ export class TasksService {
...
+  getActionItem(itemId: string): Observable<ActionItem> | undefined {
+    const actionItem = this.dataTable.find(({ id }) => id === itemId);
+    if (actionItem) {
+      return new Observable(observer => {
+        setTimeout(() => {
+          observer.next(actionItem);
+        }, 1000);
+      });
+    } else {
+      return new Observable(observer => {
```
Comment (odudycz) [inline src/app/services/tasksService/tasks.service.ts]: **NOTE:** In this case, we should probably return Observable with error (using eg. `return throwError(err);` and do proper error handling.

File: src/app/add-item/add-item.component.ts [hunk 14]
```diff
@@ -53,14 +51,16 @@ export class AddItemComponent implements OnInit {
   }
 
   private formNewActionItem(): ActionItem {
-    const projectNameValue = this.dialogForm.get('project').value;
-    const dueDateValue = this.dialogForm.get('dueDate').value;
+    const projectNameValue = this.dialogForm.get('project')!.value;
+    const dueDateValue = this.dialogForm.get('dueDate')!.value;
+    const itemId = parseInt(this.tasksService.getLastItemId()) + 1;
```
Comment (odudycz) [inline src/app/add-item/add-item.component.ts]: **SUGGESTION:** As in the end usage this `id` will be generated on the server then it would be better to assign it in the service.

File: src/app/action-items/action-items.component.ts [hunk 15]
```diff
@@ -56,8 +53,28 @@ export class ActionItemsComponent implements OnInit {
...
       });
   }
+  private getActionItem(): void {
+    this.actionItem = this.tasksService.getActionItem(this.actionItemId);
+  }
+  getQueryParams(): void {
+    this.isLoadingActionItem = true;
+    this.subscription = this.route.queryParams.subscribe(params => {
+      this.actionItemId = params.id;
+      this.getActionItem();
```
Comment (odudycz) [inline src/app/action-items/action-items.component.ts]: **SUGGESTION:** I think that it would be better to call it only if the `actionItemId` is defined. If it's not defined then just assign action item to `undefined` - the case with default route where nothing is selected.

File: karma.conf.js [hunk 16]
```diff
@@ -23,7 +23,7 @@ module.exports = function(config) {
         reporters: ['progress', 'kjhtml'],
         port: 9876,
         colors: true,
-        logLevel: config.LOG_INFO,
+        // logLevel: config.LOG_INFO,
```
Comment (odudycz) [inline karma.conf.js]: **CONCERN:** Why is it commented out?

File: src/app/action-items/action-items.component.ts [hunk 17]
```diff
@@ -60,4 +56,33 @@ export class ActionItemsComponent implements OnInit {
...
+        err => {
+          this.changePath();
+        }
+      );
+  }
+  getQueryParams(): void {
+    this.isLoadingActionItem = true;
+    this.querySubscription = this.route.queryParams.subscribe(params => {
+      this.actionItemId = params.id;
+      this.getActionItem(this.actionItemId);
```
Comment (odudycz) [inline src/app/action-items/action-items.component.ts]: This method shouldn't be call if `this.actionItemId` is not defined.

File: src/app/action-items/action-items.component.ts [hunk 18]
```diff
@@ -14,38 +15,41 @@ export interface ActionItemMapped extends ActionItem {
...
+  private querySubscription: Subscription;
 
   constructor(
     private tasksService: TasksService,
-    private daysCountService: DaysLeftToDeadlineService,
-    private matDialog: MatDialog
+    private matDialog: MatDialog,
+    private route: ActivatedRoute,
+    private router: Router,
+    private renderer: Renderer,
```
Comment (mdybich) [inline src/app/action-items/action-items.component.ts]: `renderer` is unused and can be safely deleted.

File: src/app/action-items/action-items.component.ts [hunk 19]
```diff
@@ -14,38 +15,41 @@ export interface ActionItemMapped extends ActionItem {
...
+    private elem: ElementRef
```
Comment (mdybich) [inline src/app/action-items/action-items.component.ts]: Same as above - this is not used.

File: src/app/services/tasksService/tasks.service.ts [hunk 20]
```diff
@@ -7,85 +7,132 @@ export interface ActionItem {
...
   }
-
   getAllItems(): Observable<ActionItem[]> {
     return new Observable(observer => {
       setTimeout(() => {
         observer.next(this.dataTable);
       }, 1000);
     });
   }
+  getActionItem(itemId: string | undefined): Observable<ActionItem> {
```
Comment (mdybich) [inline src/app/services/tasksService/tasks.service.ts]: You can use if statement directly in the constructor for Observable, something like:
```
getAcitonItem(id) {
  const actionItem = this.dataTable.find(...);
  
  return new Observable(observer => {
    if (actionItem) {
      observer.next(...);
    } else {
      observer.error(...);
    }
}
```

This will reduce code duplication in this function

File: src/app/item-details/item-details.component.ts [hunk 21]
```diff
@@ -1,15 +1,26 @@
...
+  @Input() currentDate: Date;
+  itemVisibility: boolean;
 
-  constructor() { }
-
-  ngOnInit() {
+  ngOnChanges() {
+    if (this.item) {
+      const itemDueDate = this.item.dueDate;
+      this.itemVisibility = typeof itemDueDate === 'string' ? false : true;
```
Comment (mdybich) [inline src/app/item-details/item-details.component.ts]: If this property is responsible for keeping information about the visibility of due date then the name of this property is wrong.

Also, you don't need to use `typeof` operator:
`this.itemVisibility = !!itemDueDate`
should be enough.

File: src/app/item-details/item-details.component.ts [hunk 22]
```diff
@@ -1,15 +1,26 @@
...
+import { DaysLeftCountedPipe } from './../pipes/daysLeftCountedPipe/days-left-counted.pipe';
+import { ActionItem } from './../services/tasksService/tasks.service';
+import { Component, Input, ChangeDetectionStrategy, OnChanges } from '@angular/core';
 
 @Component({
   selector: 'app-item-details',
   templateUrl: './item-details.component.html',
-  styleUrls: ['./item-details.component.css']
+  styleUrls: ['./item-details.component.css'],
+  providers: [DaysLeftCountedPipe],
```
Comment (mdybich) [inline src/app/item-details/item-details.component.ts]: It would be better to provide this pipe in some module rather than in the component.

File: src/app/item-details/item-details.component.html [hunk 23]
```diff
@@ -1,2 +1,49 @@
...
+            <div class="detail projectName">
+                <div>{{item.projectName}}</div>
+            </div>
+        </div>
+        <div class="center">
+            <div class="detail dueDate">
+                <div>Due Date: {{(item.dueDate | date: 'MM/dd/yyyy') || '-' }}</div>
+            </div>
+            <div class="circularDiagrams">
+                <div class="diagram positionRelative" [class.diagramsHidden]="!itemVisibility">
```
Comment (mdybich) [inline src/app/item-details/item-details.component.html]: Please use `ngIf` directive instead of adding class with display none attribute.

File: src/app/services/tasksService/tasks.service.ts [hunk 24]
```diff
@@ -7,85 +7,132 @@ export interface ActionItem {
...
       setTimeout(() => {
-        this.dataTable.push(item);
-        observer.next(item);
+        const newActionItem = {
+          ...item,
+          id: this.getLastItemId()
+        };
+        this.dataTable.push(newActionItem);
+        observer.next(newActionItem);
+        this.justAddedItem = this.getLastItem();
```
Comment (odudycz) [inline src/app/services/tasksService/tasks.service.ts]: **CONCERN:** Why do you need this assignment if you're returning this item as a result?

File: src/app/action-items/action-items.component.ts [hunk 25]
```diff
@@ -14,38 +15,41 @@ export interface ActionItemMapped extends ActionItem {
...
+export class ActionItemsComponent implements OnInit, OnDestroy {
   dataSource: ActionItemMapped[];
-  isloadingActionItems = false;
-  private currentDate: Date = new Date();
+  isLoadingActionItems = false;
+  isLoadingActionItem = false;
+  actionItemId: string | undefined;
+  actionItem: ActionItem;
+  errorMessage: string | undefined;
+  justAddedItem: ActionItem;
```
Comment (odudycz) [inline src/app/action-items/action-items.component.ts]: **FIX:** Try to prepare solution without this variable. Hint: you have all the information that you need after creation, so try to start from the code responsible for that.

File: src/app/add-item/add-item.component.ts [hunk 26]
```diff
@@ -52,15 +50,21 @@ export class AddItemComponent implements OnInit {
...
-    const newItem: ActionItem = {
-      title: this.dialogForm.get('name').value,
-      projectName: projectNameValue.name,
+  private formNewActionItem(): AddActionItem {
+    const title = this.dialogForm.get('name');
+    const projectName = this.dialogForm.get('project');
+    const dueDate = this.dialogForm.get('dueDate');
+
+    if (!title || !projectName || !dueDate) {
+      throw 'Invalid Action Item data';
```
Comment (odudycz) [inline src/app/add-item/add-item.component.ts]: **FIX:** I don't see test for that. Please make sure that you have code coverage at 100%.

File: e2e/src/displayItemDetails/displayItemDetails.e2e-spec.ts [hunk 27]
```diff
@@ -0,0 +1,34 @@
...
+import { DisplayItemDetailsPage } from './displayItemDetails.po';
+
+describe('workspace-project App', () => {
+  let page: DisplayItemDetailsPage;
+
+  beforeEach(() => {
+    page = new DisplayItemDetailsPage();
+  });
+
+  it('should display table with 8 action items', () => {
```
Comment (odudycz) [inline e2e/src/displayItemDetails/displayItemDetails.e2e-spec.ts]: **SUGGESTION:** That's minor, but the tests should be also documentation and came directly from the requirements, so it's better to name them based on the business scenarios so eg. `should display table with multiple items`. this exact number `8` is not an exact requirement. It's also not the case for the current case (as you're using mocked data) but in the real world you won't know how many items are returned from the api, so it's safer to be greater (or equal) 1.

File: e2e/protractor.conf.js [hunk 28]
```diff
@@ -8,25 +8,25 @@ const { SpecReporter } = require('jasmine-spec-reporter');
...
-  },
-  onPrepare() {
-    require('ts-node').register({
-      project: require('path').join(__dirname, './tsconfig.json')
-    });
-    jasmine.getEnv().addReporter(new SpecReporter({ spec: { displayStacktrace: true } }));
-  }
+    allScriptsTimeout: 11000,
+    specs: [
+        './src/**/addAndDisplayDetails.e2e-spec.ts'
```
Comment (odudycz) [inline e2e/protractor.conf.js]: **SUGGESTION:** Don't forget to revert this line change before merging.

File: e2e/src/addAndDisplayDetails/addAndDisplayDetails.e2e-spec.ts [hunk 29]
```diff
@@ -9,12 +9,13 @@ describe('workspace-project App', () => {
   });
 
   it('should display table with 8 action items', () => {
-    page.navigateTo();
+    page.navigateToAllActionItems();
```
Comment (odudycz) [inline e2e/src/addAndDisplayDetails/addAndDisplayDetails.e2e-spec.ts]: **FIX** Still tests are using hardcoded ids instead of pattern:
1. Create Action Item.
2. Check if it's displayed on the list.
3. Check if it's displayed in the details etc.

File: src/app/action-items/action-items.component.css [hunk 30]
```diff
@@ -146,6 +150,11 @@
     justify-content: center;
 }
 
+.fa-edit {
```
Comment (odudycz) [inline src/app/action-items/action-items.component.css]: **SUGGESTION:** It's usually better if we want to change the look of the generic class only in the single place to make a chain (nesting) of the css classes so eg. `.details. .edit-button .fa-edit`. Then we won't unexpectedly change the look in other places. It also explains easier why and what we want to change.

File: src/app/action-items/action-items.component.ts (hunk 25 above)
Comment (odudycz) [reply src/app/action-items/action-items.component.ts]: @romatowskak this is still not removed. I think that it can be safely removed as you're only assigning it and not doing anything with it after.

File: src/app/action-items/action-items.component.ts [hunk 31]
```diff
@@ -57,7 +58,36 @@ export class ActionItemsComponent implements OnInit {
...
+        item => {
+          this.isLoadingActionItem = false;
+          this.actionItem = item;
+        },
+        err => {
+          if (err.status === 404) this.errorMessage = err.statusText;
+        }
+      );
+  }
+  getQueryParams(): void {
```
Comment (odudycz) [inline src/app/action-items/action-items.component.ts]: **FIX:** This can be private
```suggestion
  private getQueryParams(): void {
```

File: src/app/action-items/action-items.component.ts [hunk 32]
```diff
@@ -57,7 +58,36 @@ export class ActionItemsComponent implements OnInit {
...
+    this.tasksService
+      .getActionItem(itemId)
+      .pipe(first())
+      .subscribe(
+        item => {
+          this.isLoadingActionItem = false;
+          this.actionItem = item;
+        },
+        err => {
+          if (err.status === 404) this.errorMessage = err.statusText;
```
Comment (odudycz) [inline src/app/action-items/action-items.component.ts]: **SUGGESTION:** It would be good to add a generic message for the other cases than `404` - eg. `Ooops! Something went wrong`.

File: src/app/action-items/action-items.component.ts [hunk 33]
```diff
@@ -57,7 +58,36 @@ export class ActionItemsComponent implements OnInit {
       .subscribe(item => {
         if (!!item) {
           this.retrieveActionItems();
+          const justAddedItemId = parseInt(this.justAddedItem.id) + 1;
```
Comment (odudycz) [inline src/app/action-items/action-items.component.ts]: **FIX:** This should be taken from the result of the dialog, so `item.id`. Doing this autoincrement won't guarantee you that this will be the id of the item generated by the server (it will only work for this case where you have dummy data generated in the service with autoincrement).

File: src/app/services/tasksService/tasks.service.ts [hunk 34]
```diff
@@ -19,73 +29,100 @@ export class TasksService {
...
+          observer.next(actionItem);
+        } else {
+          observer.error(new HttpResponse({ status: 404, statusText: 'No item with such ID found!' }));
+        }
+      }, 1000);
+    });
+  }
+  getLastItemId(): string {
+    const lastItem = this.dataTable[this.dataTable.length - 1];
+    const newItemId = parseInt(lastItem.id) + 1;
```
Comment (odudycz) [inline src/app/services/tasksService/tasks.service.ts]: **FIX:** Could you use the new id from Guid?

File: src/app/services/tasksService/tasks.service.ts [hunk 35]
```diff
@@ -19,73 +29,100 @@ export class TasksService {
       projectName: 'CASD Wilson & Lamberton Middle Schools',
       type: 'General',
       completed: '80',
-      dueDate: new Date('2019/11/17')
+      dueDate: new Date('2019/11/17'),
+      id: '1'
```
Comment (odudycz) [inline src/app/services/tasksService/tasks.service.ts]: **FIX:** Could you generate the id (and for other test items) from the guids?

File: e2e/src/addAndDisplayDetails/addAndDisplayDetails.e2e-spec.ts [hunk 36]
```diff
@@ -1,5 +1,5 @@
-import { AppPage } from './app.po';
-import { browser, by, element } from 'protractor';
+import { browser } from 'protractor';
```
Comment (mdybich) [inline e2e/src/addAndDisplayDetails/addAndDisplayDetails.e2e-spec.ts]: Unused import.

File: e2e/src/app.po.ts [hunk 37]
```diff
@@ -75,4 +75,28 @@ export class AppPage {
   getNewItem(itemName): ElementFinder {
     return element(by.cssContainingText('.title', itemName));
   }
+
+  navigateToTheFirstActionItem(): promise.Promise<string> {
```
Comment (mdybich) [inline e2e/src/app.po.ts]: Unused function.

File: e2e/src/app.po.ts [hunk 38]
```diff
@@ -75,4 +75,28 @@ export class AppPage {
...
+    return browser.get('/items?id=1');
+  }
+
+  navigateToTheSecondActionItem(): promise.Promise<string> {
```
Comment (mdybich) [inline e2e/src/app.po.ts]: Unused function.

File: src/app/action-items/action-items.component.html [hunk 39]
```diff
@@ -18,16 +18,17 @@
...
             </div>
         </div>
-        <div *ngIf="isloadingActionItems" class="spinner">
+        <div *ngIf="isLoadingActionItems" class="spinner">
             <i class="fa fa-spinner fa-spin"></i>
         </div>
-        <div class="tBody" *ngIf="!isloadingActionItems">
-            <div class="flexTable row singleItem" role="rowgroup" *ngFor="let t of dataSource">
+        <div class="tBody" *ngIf="!isLoadingActionItems">
+            <div *ngFor="let t of dataSource" id="{{ t.id }}" [routerLink]="['/items']" [queryParams]="{ id: t.id }" routerLinkActive="router-link-active" class="flexTable row singleItem" role="rowgroup">
```
Comment (mdybich) [inline src/app/action-items/action-items.component.html]: It would be better instead of `t` name this variable something like `actionItem`

File: src/app/action-items/action-items.component.html [hunk 40]
```diff
@@ -36,12 +37,18 @@
...
+                <div class="flexRow bodyCell inner" role="cell">
+                    <a><i class="fa fa-edit"></i></a>
+                </div>
             </div>
         </div>
     </div>
-    <div class="innerContainer detailsContainer">
-        <app-item-details></app-item-details>
+    <div class="innerContainer detailsContainer ">
+        <app-item-details [id]="actionItemId" [item]="actionItem" [isLoadingActionItem]="isLoadingActionItem" [currentDate]="currentDate" [errorMessage]="errorMessage">
```
Comment (mdybich) [inline src/app/action-items/action-items.component.html]: `currentDate` is not defined in the component so you are always passing undefined here. Why?

File: src/app/action-items/action-items.component.ts [hunk 41]
```diff
@@ -14,38 +16,41 @@ export interface ActionItemMapped extends ActionItem {
...
-  private currentDate: Date = new Date();
+  isLoadingActionItems = false;
+  isLoadingActionItem = false;
+  actionItemId: string | undefined;
+  actionItem: ActionItem;
+  errorMessage: string | undefined;
+  justAddedItemId: string;
+  private querySubscription: Subscription;
+  private itemRequest: Subscription;
+  public ngxScrollToDestination: string;
```
Comment (mdybich) [inline src/app/action-items/action-items.component.ts]: Unused property. Also in TS all properties are public by default so there is no need to add it explicit.

File: src/app/action-items/action-items.component.ts [hunk 42]
```diff
@@ -14,38 +16,41 @@ export interface ActionItemMapped extends ActionItem {
...
-  isloadingActionItems = false;
-  private currentDate: Date = new Date();
+  isLoadingActionItems = false;
+  isLoadingActionItem = false;
+  actionItemId: string | undefined;
+  actionItem: ActionItem;
+  errorMessage: string | undefined;
+  justAddedItemId: string;
+  private querySubscription: Subscription;
+  private itemRequest: Subscription;
```
Comment (mdybich) [inline src/app/action-items/action-items.component.ts]: `getActionItemSubscription` will be better name here.

File: src/app/action-items/action-items.component.ts [hunk 43]
```diff
@@ -14,38 +16,41 @@ export interface ActionItemMapped extends ActionItem {
...
+    private matDialog: MatDialog,
+    private route: ActivatedRoute,
+    private router: Router,
+    private elRef: ElementRef,
+    private scrollToService: ScrollToService
   ) {}
+
   ngOnInit() {
     this.retrieveActionItems();
+    this.getQueryParams();
```
Comment (mdybich) [inline src/app/action-items/action-items.component.ts]: `getQueryParams` name suggests that this function will return some query parameters. Instead, it is responsible for subscribing to the query params but it does not return anything. Please provide better name.

File: src/app/action-items/action-items.component.ts [hunk 44]
```diff
@@ -57,7 +62,49 @@ export class ActionItemsComponent implements OnInit {
       .subscribe(item => {
         if (!!item) {
           this.retrieveActionItems();
+          this.justAddedItemId = this.dataSource[this.dataSource.length - 1].id;
+          this.router.navigate(['/items'], { queryParams: { id: this.justAddedItemId } });
+          setTimeout(() => {
+            item = this.elRef.nativeElement.querySelector(`[data-id="${this.justAddedItemId}"]`);
+            this.triggerScrollTo();
+          }, 1000);
```
Comment (mdybich) [inline src/app/action-items/action-items.component.ts]: FIX: This will work because of the assumption that refreshing the action item list will take 1 second. If you go to the `tasks.service.ts` and increase the timeout value, for example:
```
  getAllItems(): Observable<ActionItem[]> {
    return new Observable(observer => {
      setTimeout(() => {
        observer.next(this.dataTable);
      }, 3000); <--------------- 3 seconds
    });
```

You will notice that scroll functionality won't work. In real-world scenario we never know how long we need to wait for the API response, so we cannot do assumption like that here.

File: src/app/item-details/item-details.component.html (hunk 23 above)
Comment (mdybich) [reply src/app/item-details/item-details.component.html]: This one is still not addressed.

File: src/app/item-details/item-details.component.ts (hunk 22 above)
Comment (mdybich) [reply src/app/item-details/item-details.component.ts]: This one is still not addressed.

File: src/app/item-details/item-details.component.ts (hunk 21 above)
Comment (mdybich) [reply src/app/item-details/item-details.component.ts]: This one is still not addressed.

File: e2e/protractor.conf.js (hunk 28 above)
Comment (odudycz) [reply e2e/protractor.conf.js]: @romatowskak what about that? still not applied

File: src/app/action-items/action-items.component.css (hunk 30 above)
Comment (odudycz) [reply src/app/action-items/action-items.component.css]: @romatowskak what about that? still not applied

File: src/app/pipes/daysLeftCountedPipe/days-left-counted.pipe.ts [hunk 45]
```diff
@@ -0,0 +1,18 @@
+import { ActionItem } from 'src/app/services/tasksService/tasks.service';
+import { Pipe, PipeTransform } from '@angular/core';
+import { DaysLeftToDeadlineService } from 'src/app/services/daysLeftToDeadlineService/days-left-to-deadline.service';
+
+@Pipe({
+  name: 'daysLeftCounted'
+})
+export class DaysLeftCountedPipe implements PipeTransform {
+  constructor(private daysCountService: DaysLeftToDeadlineService) {}
+  transform(value: ActionItem, currentDate: Date = new Date()): number | undefined {
```
Comment (odudycz) [inline src/app/pipes/daysLeftCountedPipe/days-left-counted.pipe.ts]: **FIX:** If you pass just property instead of the action item object then this pipe would be more generic and could be used for the other cases.

File: src/app/pipes/daysLeftCountedPipe/days-left-counted.pipe.ts (hunk 12 above)
Comment (odudycz) [reply src/app/pipes/daysLeftCountedPipe/days-left-counted.pipe.ts]: @romatowskak what about that? still not applied

File: src/app/item-details/item-details.component.spec.ts [hunk 46]
```diff
@@ -1,25 +1,91 @@
...
+  it('should display error message if there is no item found', () => {
+    component.item = undefined;
+    component.errorMessage = 'No item found!';
+    component.isLoadingActionItem = false;
+    fixture.detectChanges();
+    const errorMessage = fixture.debugElement.query(By.css('.selectItem.errorMessage'));
+    expect(errorMessage).toBeTruthy();
+  });
+
+  it('should display "Select Action Item.." message', () => {
```
Comment (odudycz) [inline src/app/item-details/item-details.component.spec.ts]: **FIX:** Please add `when` or `if` explanation so eg.
```suggestion
  it('should display "Select Action Item.." message if no action item was selected', () => {
```

File: src/app/item-details/item-details.component.ts [hunk 47]
```diff
@@ -1,15 +1,25 @@
...
+  changeDetection: ChangeDetectionStrategy.OnPush
 })
-export class ItemDetailsComponent implements OnInit {
+export class ItemDetailsComponent implements OnChanges {
+  @Input() item: ActionItem | undefined;
+  @Input() id: string;
+  @Input() errorMessage: string | undefined;
```
Comment (odudycz) [inline src/app/item-details/item-details.component.ts]:
```suggestion
  @Input() errorMessage?: string;
```
